accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+#endif
+```
+
+**Located in xamarin/xamarin-macios repo**
+```c#
+// Sample implementation in System.Net.Http.Private.dll build as part of Xamarin.iOS
+public class HttpClientHandler
+{
+  public HttpClientHandler ()
+  {
+  }
+
+  public virtual bool SupportsProxy => nativecall ("has-proxy") as bool;
+
+  public int MaxAutomaticRedirections {
+    get => 100;
+    set => throw PlatformNotSupportedException();
+  }
+
+  // The rest of API contract follows
+```
+
+**Disadvantages**
+
+* Single type assembly is against our design practices
+* A circular dependency between System.Net.Http.Private.dll and System.Net.Http.dll might exist in optional components
+* API contract would live in different assembly than implementation
+* Reliance on more dotnet publish to figure out right implementation assembly
+* Adding public member to HttpClientHandler would be a breaking change
+* HttpClientHandler acts as a wrapper in most SDKs and the code would have to be duplicated
+* Platform specific implementation would need to write their own copy of proxy logic
+
+**Conclusion**
+
+This is very light-weight solution which pushes a lot of responsibility to optional components and build system. The possibility of including RID specific version of System.Net.Http.Private.dll inside Xamarin.iOS.nupkg needs to be confirmed.
+
+### Proposal #4: Add HttpClientHandler type forwarder to SDKs assemblies
+
+.NET builds ref version System.Net.Http for all optional components inside dotnet/runtime by using references to copies of optional component reference assemblies. The implementation then type forwards to optional components implementation assemblies.
+
+We could also use have only shared ref version of HttpClientHandler but then every optional SDKs would have to create their own copy of proxy logic as described in **Proposal #2**.
+
+**Located in dotnet/runtime repo**
+```c#
+// System.Net.Http.dll - ref for TFM net5.0-ios
+public class HttpClientHandler
+{
+  public HttpClientHandler ()
+  {
+  }
+
+  public long MaxInputInMemory { get; set; }
+
+  public NSUrlSessionHandlerTrustOverrideCallback TrustOverride { get; set; }
+  
+  // More unique or common properties to follow
+}
+```
+
+```c#
+// System.Net.Http.dll
+
+#if iOS
+// The reference for the type comes from Xamarin.iOS.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#elif (android)
+// The reference for the type comes from Xamarin.Android.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#else if ..
+
+#endif
+```
+
+**Disadvantages**
+
+* dotnet/runtime will need to include reference assemblies of all optional components to be able to build ref and impl assemblies.
+* Public HttpClientHandler API shape might be different between TFM as it’s controlled by optional components
+* Updating HttpClientHandler public surface for optional component will require build a new version of dotnet/runtime
```

Comment:
How is that gonna be tested ?

What happen when a new native API is added and support must be added back to `dotnet` and backported to current releases (in sync with other changes made in the SDK) ?

It might require a newer Xcode (than the one currently used) and that can bring its own fun (like the issue we had with Xcode 11.4).

Reply:
> How is that gonna be tested ?

Same as other proposals depending on implementation in other repo.